README
Working with RDF data is cumbersome, unless you define namespaces and prefixes
for abbreviation. Defining namespaces, however, is also cumbersome because you
need to repeat this step in every environment. This is silly because in most
cases we use the same URI namespaces anyway. Some popular examples:

    @prefix owl: <http://www.w3.org/2002/07/owl#> .   (Turtle/N3)

    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>   (SPARQL)

    xmlns:foaf="http://xmlns.com/foaf/0.1/"   (RDF/XML)

Fortunately there is http://prefix.cc, a registry where people can vote the most
popular URI namespaces. It is best practice to use the same prefixes as defined
there. The Perl module RDF::NS contains regular snapshots from http://prefix.cc
for easy reuse without having to download them in every application.

In short, you can get a full prefix-to-namespace mapping as blessed hash with

    RDF::NS->new('20130402')

The only argument to the constructor is the version number, that is equivalent
to the date of a particular snapshot. The changelog in the file Changes lists
all snapshots with their modifications. RDF::NS provides some handy methods,
for instance you can use method-calls to create full URIs:

    $ns->rdf_type; # returns "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"

The subclass RDF::NS::Trine, shipped with RDF::NS returns blessed URI objects
instead of plain strings, as expected by RDF::Trine. Please have a look at the
documentation for more information.

The source code of RDF::NS is available at https://github.com/nichtich/RDF-NS/
with releases at http://search.cpan.org/dist/RDF-NS/.


Changes
{{$NEXT}}

20230619 2023-06-19T08:56:28Z
 - 2971 prefixes
 - added:   accid acp adf aec3po aff afm afr ags alg ammo ams apb apods
            archivo arena arp asf asgs asio aspect atcc atd atm atts baf
            bag2 bao bau bcfowl bcom bdg bdsubj beer beo besluitvor biocrm
            biogrid biolink birthdate bitl bkb bld bldont ble bleadapter
            bmp bop bp3 bperson bpo brick brot bs bsdd bsh ca capes caso
            cbb cbs cbv cci ccomid ccp ccr ccsla ceox cerealstoo cfrl cgo
            ch chameo chear check chemrof chemsci chess ci ciao cido cinema
            citedcat coda cog comp compub conll conllu consolid constant
            contax contry coos coreo country county covido coy cpc cpg
            crmeh crminf crmsci cska cso cto ctrl cts cur cvb cwmo d2s daap
            databus datagc dave dbd dbm dbms dbonto dbowl dbpg dby dcatno
            dcb dentsci devuan dfc dfcb dick diso djo dm dmo dmp docam doce
            dom donto dprov dpv dqvno drk dsd dsi ea earth eb ebg eccf ecfo
            edu edupro eep ei2a encargado epcis eppl epplan eppo eproc
            eproc2 eqp era ermrk esagen esagm esautob esbici esc esconv
            esempleo esgs esservicio estraf eupont eurio ewg exo experts
            express extech faas fdof fdp feed fel fernanda fibo film fnom
            foio folio fpr freq fso fx gas gax gco gcon gdprov gdprtext
            generiek genre geodcat geofabric ggbn ghga gleio gmo gnaf
            goavoc godaddy gom gql grscicoll gufo gx hctl hdgi hico hmas
            hni hola hops hosp hpont hqdm hqdmontol hu hyr i18n iaco iaph
            ibeacon icon iddo idot idpo idsc ii imbor imjv inchikey
            ingredient interop io iospress iottta isoadr istex itcrdf itops
            jsonld jsonschema jup jur karstlink kdsf kgi kko knows koko
            kokot kpd ksam la lado lbds lbdserver lblodlg lcnaf ld ldc ldes
            ldpsc lds lesa lexicog lg linkedart linkml loci lol losp lprov
            lrcommon lsqr luigiusai lv2 m3lite m4i m8g mag magmardl
            magmauser malaka manto marc marcgt maroc matmine matrycs matvoc
            mbkeys mccv mdcs melding memorix mesh2021 mgv mindat misp mltd
            mmms mnx mobiliteit modsci mondo motogp movieo mpg123 mr mrk
            mrt mus mwapi nalt nanopub nas nbo ncbi nen2660 nno noise nom
            nomo notify nprl nsd nsg ntp nutscode oav obws occ ocds oco
            ocsw odf ods oeo oeso ofn ofo ogcf ogham oghamonto ogura oidc
            olias omim omnlc omop one ontobras ontologia ontouml oop oplben
            oplfeat oplsoft opltw ordo osdu osmm osmnode osmo osmrel osmt
            osmway osys otl oup p2po paam paf peco perscido persoon pgo
            pgxo pham phto phy pineapple pit pnv poke powla ppeer pplan pq
            prismdb probont property proton pubchem qk qkdv quid quran raad
            radar rami rank raum rb rbdoc rc rcgs rdaao rdaed rdaeo rdaep
            rdaid rdaim rdaio rdamat rdamd rdamo rdan rdand rdano rdap
            rdapath rdapd rdapol rdasca rdasource rdat rdatask rdatb rdatd
            rdato rdaut rdawd rdawo rdaxd rdaxo react refexo reg relation
            respond rico risk rism rkd rl rmlt rnce roar robotarm roc
            rofchrda roffgrda rofitrda rofsfrda roh rpg rro rsctx rsp rsx
            rtedurp rto s223 s4agri s4city s4ee s4ener s4syst samian
            saref4envi scho schoi sds seasd semsur sense setl sfl she shema
            signify sirext skoslex skosm skosthes slm smithy snomedct soall
            sohu sohua sohuc sohud sohuea sohueccont sohuecons sohues sohuj
            sohukd sohur sohut sohutr soma somaae somab somac somael somaem
            somaen somaeq somaf somaind somains somaio somais somamic
            somamin soman somanr somaoc somapa somapl somarock somarocki
            somas somaw soph sophatmo sophatmoc sophatmofo sophatmofr
            sophatmol sophatmopc sophatmops sophatmos sophatmot sophatmow
            sophatmowm sophb sophcm sophcr sophcy sophec sophei sophel
            sophen sophfd sophfi sophft sophg sophgf sophgg sophgs sophgt
            sophgv sophhe sophhy sophm sopho sophoc sophod sophpc sophr
            sophso sophst sophsy sophsyc sophw sophwn soproc soprocc
            soprocp soprocsc soprocw soprop sopropb sopropcap sopropcat
            sopropcha sopropche sopropcon sopropcou sopropdife sopropdifu
            sopropdr soprope sopropef sopropfr sopropfu sopropi sopropm
            sopropmf sopropo sopropp sopropq sopropr soprops sopropsdir
            sopropsdis sopropsh sopropsl sopropsm sopropsp sopropst sopropt
            soproptf soproptg sopropti sorea soreaa soreaab soreaabl
            soreaah soreaas soreaaw soreabb soreac soreacz soreaer soreahb
            soreal soreala sorealc sorealf sorealg sorealo sorealp sorealt
            sorealv soreao soreaofe soreaofl sorear soreas sorel sorelch
            sorelcl sorelh sorelm sorelph sorelpr sorelsc sorelsp sorelt
            sorep sorepdf sorepdm sorepdp sorepds sorepdsa sorepdsg
            sorepdsr sorepdsv sorepm sorepmf sorepmfo sorepmg sorepmo
            sorepmso sorepmst soreps sorepsc sorepscd sorepsd sorepsf
            sorepsg3 sorepsl sorepsme sorepsmo sorepsp sorepsrs sorepsu
            sorept soreptd sorepts sost sostb sostc sostdp sostef sostf
            sosto sostp sostrb sostrc sostre sostrg sostri sostro sostrr
            sostrt sostsb sostsc sostsl sostso sostsp sostss sostst sostsy
            sosttc sosttf sosttg sostth sostti sostv sou spatialf spec
            spvqa srmo srr srt srtun srv ssnx stat stencila stirdata stix
            stx sty summa sure swa sweet taxref taxrefprop tb tci tempo
            terms tern textgrid threat thub toaru trak trao tree trek tro
            tso tsso ttrpg uberon ucum ufo uimo ul ultragaz univ url value
            vcs veelana veo version vhbieo vir vl vlueprint vntourism
            vocals vph vr w3geo w3id waa walmart wasa wdno wds wdtn weki

update.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::Simple qw(mirror);
use RDF::NS;
use File::Temp;
use Text::Wrap;
$Text::Wrap::unexpand = 0;    # don't insert tabs

# make sure, git repository is clean
my $dirty = `git status --porcelain`;
die "git repository is dirty\n" if $dirty;

# get current version distribution
my $dist = do { local ( @ARGV, $/ ) = 'lib/RDF/NS.pm'; <> };
my $cur_version = $1 if $dist =~ /^our \$VERSION\s*=\s*'([^']+)'/m;
$cur_version or die 'current version not found in lib/RDF/NS.pm';

# get current prefixes
my $cur = RDF::NS->LOAD( "share/prefix.cc", warn => 1 );
die "share/prefix.cc is empty" unless %$cur;

# get new current datestamp
my @t           = gmtime;
my $new_version = sprintf '%4d%02d%02d', $t[5] + 1900, $t[4] + 1, $t[3];
my $new_date    = sprintf '%4d-%02d%-%02d', $t[5] + 1900, $t[4] + 1, $t[3];
die "$new_version is not new" if $new_version eq $cur_version;

# download new prefixes
my $tmp = File::Temp->new->filename;
my $url = "http://prefix.cc/popular/all.file.txt";
mirror( $url, $tmp ) or die "Failed to load $url";
my $new = RDF::NS->LOAD( $tmp, warn => 1 );

# lock all prefixes pointing to example.* TLD URLs
for ( grep { $new->{$_} =~ qr{^https?://example.\w} } keys %$new ) {
    delete $new->{$_} if $new->{$_} ne $cur->{$_};
}

my $diff = $new->UPDATE( "share/prefix.cc", $new_version );

foreach my $change (qw(create delete update)) {
    my $prefixes = $diff->{$change} or next;
    foreach my $prefix (@$prefixes) {
        if ( $change eq 'create' ) {
            printf "+ $prefix %s\n", $new->URI($prefix);
        }
        elsif ( $change eq 'delete' ) {
            printf "- $prefix %s\n", $cur->URI($prefix);
        }
        else {
            printf "- $prefix %s\n", $cur->URI($prefix);
            printf "+ $prefix %s\n", $new->URI($prefix);
        }
    }
}

my (@log) = ( '{{$NEXT}}', ' - ' . $new->COUNT . " prefixes" );

sub prefix_list {
    my ( $label, $list ) = @_;
    return wrap( " - $label ", " " x 12, join ' ', @$list );
}

push @log, prefix_list( "added:  ", $diff->{create} ) if @{ $diff->{create} };
push @log, prefix_list( "removed:", $diff->{delete} ) if @{ $diff->{delete} };
push @log, prefix_list( "changed:", $diff->{update} ) if @{ $diff->{update} };

my @files = qw(
  lib/App/rdfns.pm
  lib/RDF/NS.pm
  lib/RDF/NS/Trine.pm
  lib/RDF/NS/URIS.pm
  lib/RDF/SN.pm
  README.md
);

foreach my $file (@files) {
    print "$cur_version => $new_version in $file\n";
    local ( $^I, @ARGV ) = ( '.bak', $file );
    while (<>) {
        s/$cur_version/$new_version/ig;
        print;
    }
}

my $msg = `git log --pretty=format:" - %s" $cur_version..`;

do {
    print "prepend modifications to Changes\n";
    local ( $^I, @ARGV ) = ( '.bak', 'Changes' );
    my $line = 0;
    while (<>) {
        if ( !$line++ ) {    # prepend
            print join '', map { "$_\n" } @log;
            print "\n$msg" if $msg;
            print "\n\n";
        }
        print;
    }
};

print `git add -u`;
print `git commit -m "update to $new_version"`;

1;


lib/RDF/SN.pm
package RDF::SN;
use v5.10;
use strict;
use warnings;

our $VERSION = '20160409';

use RDF::NS;
use Scalar::Util qw(blessed);

sub new {
    my ($class, $ns) = @_;

    unless (blessed $ns) {
        $ns = $ns ? RDF::NS->new($ns) : RDF::NS->new;
    }

    my $self = bless { }, $class;
    
    while ( my ($prefix, $namespace) = each %$ns ) {
        my $has = $self->{$namespace};
        if (!$has || (length($has) > length($prefix))
                  || (length($has) == length($prefix) and $has ge $prefix)
        ) {
            $self->{$namespace} = $prefix;
        }
    }

    $self;
}

sub qname {
    my ($self, $uri) = @_;

    if ($self->{$uri}) {
        return wantarray ? ($self->{$uri}, '') : $self->{$uri}.':';
    }

    # regexpes copied from RDF::Trine::Node::Resource
    our $r_PN_CHARS_BASE ||= qr/([A-Z]|[a-z]|[\x{00C0}-\x{00D6}]|[\x{00D8}-\x{00F6}]|[\x{00F8}-\x{02FF}]|[\x{0370}-\x{037D}]|[\x{037F}-\x{1FFF}]|[\x{200C}-\x{200D}]|[\x{2070}-\x{218F}]|[\x{2C00}-\x{2FEF}]|[\x{3001}-\x{D7FF}]|[\x{F900}-\x{FDCF}]|[\x{FDF0}-\x{FFFD}]|[\x{10000}-\x{EFFFF}])/;
    our $r_PN_CHARS_U    ||= qr/(_|${r_PN_CHARS_BASE})/;
    our $r_PN_CHARS      ||= qr/${r_PN_CHARS_U}|-|[0-9]|\x{00B7}|[\x{0300}-\x{036F}]|[\x{203F}-\x{2040}]/;
    our $r_PN_LOCAL      ||= qr/((${r_PN_CHARS_U})((${r_PN_CHARS}|[.])*${r_PN_CHARS})?)/;

    if ($uri =~ m/${r_PN_LOCAL}$/) {
        my $ln = $1;
        my $ns = substr($uri, 0, length($uri)-length($ln));
        if ($self->{$ns}) {
            return(wantarray ? ($self->{$ns},$ln) : $self->{$ns}.':'.$ln);
        }
    }

    return;
}

sub qname_ {
    if(wantarray) {
        return $_[0]->qname($_[1]);
    } else {
        return join '_', $_[0]->qname($_[1]);
    }
}

1;
__END__

=head1 NAME

RDF::SN - Short names for URIs with prefixes from prefix.cc

=head1 SYNOPSIS

  use RDF::SN;
  $abbrev = RDF::SN->new('20160409');
  $abbrev->qname('http://www.w3.org/2000/01/rdf-schema#type'); # rdfs:type

=head1 DESCRIPTION

This module supports abbreviating URIs as short names (aka qualified names), so
its the counterpart of L<RDF::NS>.

=head2 new( [ $ns ] )

Create a lookup hash from a mapping hash of namespace URIs to prefixes
(L<RDF::NS>). If multiple prefixes exist, the shortest is used. If multiple
prefixes with same length exist, the first in alphabetical order is used.

=head2 qname( $uri )

Returns a prefix and local name (as list in list context, concatenated by C<:>
in scalar context) if the URI can be abbreviated with given namespaces.

=encoding utf8

=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2013- by Jakob Voß.
 
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/App/rdfns.pm
package App::rdfns;
use v5.10;
use strict;
use warnings;

use RDF::NS;

our $VERSION = '20160409';

sub new {
    bless {}, shift;
}

sub run {
    my ($self, @ARGV) = @_;
    my $format = '';

    return $self->usage if !@ARGV or $ARGV[0] =~ /^(-[?h]|--help)$/;
    return $self->version if $ARGV[0] =~ /^(-v|--version)$/;
    return $self->dates if $ARGV[0] eq '--dates';

    my $ns = RDF::NS->new;
    my $sn = $ns->REVERSE;

    foreach my $a (@ARGV) {
        if ( $a =~ /^([0-9]{8})$/ ) {
            $ns = RDF::NS->new($a);
            $sn = $ns->REVERSE;
            next;
        }
        if ( $a =~ qr{^https?://} ) {
            my $qname = $sn->qname($a);
            if ($qname) {
                $qname =~ s/:$//;
                say $qname;
            }
        } elsif ( $a =~ /:/ ) {
            print map { $ns->URI($_)."\n" } split(/[|, ]+/, $a);
        } elsif ( $a =~ s/\.([^.]+)$// ) {
            my $f = $1;
            if ( $f eq 'prefix' ) {
               print map { "$_\n" if defined $_ } map {
                   $sn->{$_}
               } $ns->FORMAT( $format, $a );
               next;
            } elsif ( $f =~ $RDF::NS::FORMATS ) {
                $format = $f;
            } else {
                print STDERR "Unknown format: $f\n";
            }
        }
        if ( lc($format) eq 'json' ) {
            say join ",\n", $ns->FORMAT( $format, $a );
        } else {
            say $_ for $ns->FORMAT( $format, $a );
        }
    }
}

sub usage {
    print <<'USAGE';
USAGE: rdfns { [YYYYMMDD] ( <prefix[es]>[.format] | prefix:name | URL ) }+

  formats: txt, sparql, ttl, n3, xmlns, json, beacon, prefix
  options: --help | --version | --dates
 
  examples:
    rdfns 20111102 foaf,owl.ttl
    rdfns foaf.xmlns foaf.n3
    rdfns rdfs:seeAlso
    rdfns http://www.w3.org/2003/01/geo/wgs84_pos#
    rdfns http://purl.org/dc/elements/1.1/title
    rdfns wgs.prefix
USAGE
    0;
}

sub version {
    print $RDF::NS::VERSION . "\n";
    0;
}

sub dates {
    my $fh = RDF::NS->new->DATA;
    my $date = ''; 
    foreach (<$fh>) {
        chomp;
        next if /^#/;
        my @fields = split "\t", $_;
        if ($fields[2] ne $date) {
            say $date=$fields[2];
        }
    } 
    0;
}

1;
__END__

=head1 NAME

App::rdfns - quickly get common URI namespaces

=head1 SEE ALSO

This module implements the command line client L<rdfns>.

=encoding utf8

=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2013- by Jakob Voß.
 
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/RDF/NS/URIS.pm
package RDF::NS::URIS;
use v5.10;
use strict;
use warnings;

our $VERSION = '20160409';

use URI;
use base 'RDF::NS';

sub GET {
    URI->new($_[1]);
}

1;
__END__

=head1 NAME

RDF::NS::URIS - Popular RDF namespace prefixes from prefix.cc as URI objects

=head1 SYNOPSIS

  use RDF::NS::URIS;
  use constant NS => RDF::NS::URIS->new('20160409');

  NS->foaf_Person;             # an URI object
  NS->uri('foaf:Person);       # same
  NS->foaf_Person->as_string;  # http://xmlns.com/foaf/0.1/Person

=head1 DESCRIPTION

RDF::NS::URIS works like L<RDF::NS> but it returns instances of L<URI> instead
of plain strings. You must have installed module L<URI> to use this package.

=head1 SEE ALSO

L<URI::NamespaceMap>

=encoding utf8

=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2013- by Jakob Voß.
 
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/RDF/NS/Trine.pm
package RDF::NS::Trine;
use v5.10;
use strict;
use warnings;

our $VERSION = '20160409';

use RDF::Trine::Node::Resource;
use RDF::Trine::Node::Blank;

use base 'RDF::NS';

sub GET {
    RDF::Trine::Node::Resource->new($_[1]);
}

sub BLANK {
    my $id = ($_[1] =~ /^_(:(.+))$/ ? $2 : undef);
    return RDF::Trine::Node::Blank->new( $id );
}

1;
__END__

=head1 NAME

RDF::NS::Trine - Popular RDF namespace prefixes from prefix.cc as RDF::Trine nodes

=head1 SYNOPSIS

  use RDF::NS::Trine;
  use constant NS => RDF::NS::Trine->new('20160409');

  NS->foaf_Person;        # iri('http://xmlns.com/foaf/0.1/Person')
  NS->uri('foaf:Person);  #  same RDF::Trine::Node::Resource
  NS->foaf_Person->uri;   # http://xmlns.com/foaf/0.1/Person

  NS->_;                  # RDF::Trine::Node::Blank
  NS->_abc;               # blank node with id 'abc'
  NS->uri('_:abc');       # same

=head1 DESCRIPTION

RDF::NS::Trine works like L<RDF::NS> but it returns instances of
L<RDF::Trine::Node::Resource> (or L<RDF::Trine::Node::Blank>) instead of
strings.

Before using this module, make sure to install L<RDF::Trine>, which is not
installed automatically together with L<RDF::NS>!

=head1 ADDITIONAL METHODS

=head2 BLANK ( [ $short ] )

Returns a new L<RDF::Trine::Node::Blank>.

=encoding utf8

=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2013- by Jakob Voß.
 
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/RDF/NS.pm
package RDF::NS;
use v5.10;
use strict;
use warnings;

our $VERSION = '20181102';

use Scalar::Util qw(blessed reftype);
use File::ShareDir;
use Carp;
use RDF::SN;

our $AUTOLOAD;
our $FORMATS = qr/ttl|n(otation)?3|sparql|xmlns|txt|beacon|json/;

our $DATE_REGEXP = qr/^([0-9]{4})-?([0-9][0-9])-?([0-9][0-9])$/;

sub new {
    my $class = ref( $_[0] ) ? ref(shift) : shift;
    my $from  = @_ % 2       ? shift      : 1;
    my %options = @_;
    my $at      = $options{at} || 'any';
    my $warn    = $options{'warn'};
    $from = $options{from} if exists $options{from};
    $from = 'any' if !$from or $from eq 1;

    if ( ( ref($from) || '' ) eq 'HASH' ) {
        my $self = bless $from, $class;
        foreach my $prefix ( keys %$self ) {
            unless ( $self->SET( $prefix => $self->{$prefix}, $warn ) ) {
                delete $self->{$prefix};
            }
        }
        return $self;
    }

    if ( $from =~ $DATE_REGEXP ) {
        $at   = "$1$2$3";
        $from = 'any';
    }
    elsif ( $at =~ $DATE_REGEXP ) {
        $at = "$1$2$3";
    }
    elsif ( $at !~ 'any' ) {
        croak "RDF::NS expects 'any', '1' or a date as YYYY-MM-DD";
    }

    my $self = bless {}, $class;
    my $fh = $self->DATA($from);
    foreach (<$fh>) {
        chomp;
        next if /^#/;
        my ( $prefix, $namespace, $date ) = split "\t", $_;
        last if $date and $at ne 'any' and $date > $at;

        $self->SET( $prefix => $namespace, $warn );
    }
    close($fh);

    $self;
}

sub DATA {    # TODO: document
    my ( $self, $from ) = @_;
    $from = File::ShareDir::dist_file( 'RDF-NS', "prefix.cc" )
      if ( $from // 'any' ) eq 'any';
    croak "prefix file or date not found: $from"
      unless -f $from;
    open( my $fh, '<', $from ) or croak "failed to open $from";
    $fh;
}

sub SET {
    my ( $self, $prefix, $namespace, $warn ) = @_;

    if ( $prefix =~ /^(isa|can|new|uri)$/ ) {
        carp "Cannot support prefix '$prefix'" if $warn;
    }
    elsif ( $prefix =~ /^[a-z][a-z0-9]*$/ ) {
        if ( $namespace =~ /^[a-z][a-z0-9]*:[^"<>]*$/ ) {
            $self->{$prefix} = $namespace;
            return 1;
        }
        elsif ($warn) {
            carp "Skipping invalid $prefix namespace $namespace";
        }
    }
    elsif ($warn) {
        carp "Skipping unusual prefix '$prefix'";
    }

    return;
}

*LOAD = *new;

sub COUNT {
    scalar keys %{ $_[0] };
}

sub FORMAT {
    my $self = shift;
    my $format = shift || "";
    $format = 'TTL' if $format =~ /^n(otation)?3$/i;
    if ( lc($format) =~ $FORMATS ) {
        $format = uc($format);
        $self->$format(@_);
    }
    elsif ( $format eq "" ) {
        $self->MAP( sub { $self->{$_} }, @_ );
    }
}

sub PREFIX {
    my ( $self, $uri ) = @_;
    foreach my $prefix ( sort keys %$self ) {
        return $prefix if $uri eq $self->{$prefix};
    }
    return;
}

sub PREFIXES {
    my ( $self, $uri ) = @_;
    my @prefixes;
    while ( my ( $prefix, $namespace ) = each %$self ) {
        push @prefixes, $prefix if $uri eq $namespace;
    }
    return ( sort(@prefixes) );
}

sub REVERSE {
    RDF::SN->new( $_[0] );
}

sub TTL {
    my $self = shift;
    $self->MAP( sub { "\@prefix $_: <" . $self->{$_} . "> ." }, @_ );
}

sub SPARQL {
    my $self = shift;
    $self->MAP( sub { "PREFIX $_: <" . $self->{$_} . ">" }, @_ );
}

sub XMLNS {
    my $self = shift;
    $self->MAP( sub { "xmlns:$_=\"" . $self->{$_} . "\"" }, @_ );
}

sub TXT {
    my $self = shift;
    $self->MAP( sub { "$_\t" . $self->{$_} }, @_ );
}

sub JSON {
    my $self = shift;
    $self->MAP( sub { "\"$_\": \"" . $self->{$_} . "\"" }, @_ );
}

sub BEACON {
    my $self = shift;
    $self->MAP( sub { "#PREFIX: " . $self->{$_} }, @_ );
}

sub SELECT {
    my $self = shift;
    $self->MAP( sub { $_ => $self->{$_} }, @_ );
}

# functional programming rulez!
sub MAP {
    my $self = shift;
    my $code = shift;
    my @ns =
      @_
      ? ( grep { $self->{$_} } map { split /[|, ]+/ } @_ )
      : keys %$self;
    if (wantarray) {
        return map { $code->() } sort @ns;
    }
    else {
        local $_ = $ns[0];
        return $code->();
    }
}

sub GET {
    $_[1];
}

sub BLANK {
}

*URI = *uri;

sub uri {
    my $self = shift;
    return $1 if $_[0] =~ /^<([a-zA-Z][a-zA-Z+.-]*:.+)>$/;
    return $self->BLANK( $_[0] ) if $_[0] =~ /^_(:.*)?$/;
    return unless shift =~ /^([a-z][a-z0-9]*)?([:_]([^:]+))?$/;
    my $ns = $self->{ defined $1 ? $1 : '' };
    return unless defined $ns;
    return $self->GET($ns) unless $3;
    return $self->GET( $ns . $3 );
}

sub AUTOLOAD {
    my $self = shift;
    return unless $AUTOLOAD =~ /^.*::([a-z][a-z0-9]*)?(_([^:]+)?)?$/;
    return $self->BLANK( defined $3 ? "_:$3" : '_' ) unless $1;
    my $ns = $self->{$1} or return;
    my $local = defined $3 ? $3 : shift;
    return $self->GET($ns) unless defined $local;
    return $self->GET( $ns . $local );
}

sub UPDATE {
    my ( $self, $file, $date ) = @_;

    croak "RDF::NS expects a date as YYYY-MM-DD"
      unless $date and $date =~ $DATE_REGEXP;
    $date = "$1$2$3";

    my $old = RDF::NS->new($file);
    my ( @create, @update, @delete );

    open( my $fh, '>>', $file ) or croak "failed to open $file";
    my @lines;

    while ( my ( $prefix, $namespace ) = each %$self ) {
        if ( !exists $old->{$prefix} ) {
            push @create, $prefix;
        }
        elsif ( $old->{$prefix} ne $namespace ) {
            push @update, $prefix;
        }
        else {
            next;
        }
        push @lines, "$prefix\t$namespace";
    }
    while ( my ( $prefix, $namespace ) = each %$old ) {
        if ( !exists $self->{$prefix} ) {
            push @delete, $prefix;
        }
    }

    print $fh "$_\t$date\n" for sort @lines;
    close $fh;

    return {
        create => [ sort @create ],
        update => [ sort @update ],
        delete => [ sort @delete ],
    };
}

1;
__END__

=head1 NAME

RDF::NS - Just use popular RDF namespace prefixes from prefix.cc

=begin markdown

# STATUS

[![Build Status](https://travis-ci.org/nichtich/RDF-NS.png)](https://travis-ci.org/nichtich/RDF-NS)
[![Coverage Status](https://coveralls.io/repos/nichtich/RDF-NS/badge.svg?branch=master)](https://coveralls.io/r/nichtich/RDF-NS?branch=master)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/RDF-NS.png)](http://cpants.cpanauthors.org/dist/RDF-NS)

=end markdown

=head1 SYNOPSIS

  use RDF::NS '20181102';              # check at compile time
  my $ns = RDF::NS->new('20181102');   # check at runtime

  $ns->foaf;               # http://xmlns.com/foaf/0.1/
  $ns->foaf_Person;        # http://xmlns.com/foaf/0.1/Person
  $ns->foaf('Person');     # http://xmlns.com/foaf/0.1/Person
  $ns->uri('foaf:Person'); # http://xmlns.com/foaf/0.1/Person

  use RDF::NS;             # get rid if typing '$' by defining a constant
  use constant NS => RDF::NS->new('20111208');
  NS->foaf_Person;         # http://xmlns.com/foaf/0.1/Person

  $ns->SPAQRL('foaf');     # PREFIX foaf: <http://xmlns.com/foaf/0.1/>
  $ns->TTL('foaf');        # @prefix foaf: <http://xmlns.com/foaf/0.1/> .
  $ns->XMLNS('foaf');      # xmlns:foaf="http://xmlns.com/foaf/0.1/"

  # load your own mapping from a file
  $ns = RDF::NS->new("mapping.txt");

  # select particular mappings
  %map = $ns->SELECT('rdf,dc,foaf');
  $uri = $ns->SELECT('foo|bar|doz'); # returns first existing namespace

  # instances of RDF::NS are just blessed hash references
  $ns->{'foaf'};           # http://xmlns.com/foaf/0.1/
  bless { foaf => 'http://xmlns.com/foaf/0.1/' }, 'RDF::NS';
  print (scalar keys %$ns) . "prefixes\n";
  $ns->COUNT;              # also returns the number of prefixes

=head1 DESCRIPTION

Hardcoding URI namespaces and prefixes for RDF applications is neither fun nor
maintainable.  In the end we all use more or less the same prefix definitions,
as collected at L<http://prefix.cc>. This module includes all these prefixes as
defined at specific snapshots in time. These snapshots correspond to version
numbers of this module. By selecting particular versions, you make sure that
changes at prefix.cc won't affect your programs.

The command line client L<rdfns> is installed automatically with this module:

  $ rdfns rdf,foaf.ttl
  @prefix foaf: <http://xmlns.com/foaf/0.1/> .
  @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

This module does not require L<RDF::Trine>, which is recommended nevertheless.
(at least version 0.140).  If you prefer RDF::NS to return instances of 
L<RDF::Trine::Node::Resource> instead of plain strings, use L<RDF::NS::Trine>.
L<RDF::NS::URIS> is a similar module that returns instances of L<URI>.

The code repository of this module contains an
L<update script|https://github.com/nichtich/RDF-NS/blob/master/update.pl>
to download the current prefix-namespace mappings from L<http://prefix.cc>.

=head1 GENERAL METHODS

In most cases you only need the following lowercase methods.

=head2 new ( [ $file_or_date ] [ %options ] )

Create a new namespace mapping from a selected file, date, or hash reference.
The special string C<"any"> or the value C<1> can be used to get the newest
mapping, but you should better select a specific version, as mappings can
change, violating backwards compatibility.  Supported options include C<warn>
to enable warnings and C<at> to specify a date. 

=head2 "I<prefix>"

Returns the namespace for I<prefix> if namespace prefix is defined. For
instance C<< $ns->foaf >> returns C<http://xmlns.com/foaf/0.1/>.

=head2 "I<prefix_name>"

Returns the namespace plus local name, if namespace prefix is defined. For
instance C<< $ns->foaf_Person >> returns C<http://xmlns.com/foaf/0.1/Person>.

=head2 uri ( $short | "<$URI>" )

Expand a prefixed URI, such as C<foaf:Person> or C<foaf_Person>. Alternatively 
you can expand prefixed URIs with method calls, such as C<< $ns->foaf_Person >>.
If you pass an URI wrapped in C<E<lt>> and C<E<gt>>, it will not be expanded
but returned as given.

=head1 SERIALIZATION METHODS

=head2 TTL ( prefix[es] )

Returns a Turtle/Notation3 C<@prefix> definition or a list of such definitions
in list context. Prefixes can be passed as single arguments or separated by
commas, vertical bars, and spaces.

=head2 SPARQL ( prefix[es] )

Returns a SPARQL PREFIX definition or a list of such definitions in list
context. Prefixes can be passed as single arguments or separated by commas,
vertical bars, and spaces.

=head2 XMLNS ( prefix[es] )

Returns an XML namespace declaration or a list of such declarations in list
context. Prefixes can be passed as single arguments or separated by commas,
vertical bars, and spaces.

=head2 TXT ( prefix[es] )

Returns a list of tabular-separated prefix-namespace-mappings.

=head2 BEACON ( prefix[es] )

Returns a list of BEACON format prefix definitions (not including prefixes).

=head1 LOOKUP METHODS

=head2 PREFIX ( $uri )

Get a prefix of a namespace URI, if it is defined. This method does a reverse
lookup which is less performant than the other direction. If multiple prefixes
are defined, the first in sorted order is returned. If you need to call this
method frequently and with deterministic response, better create a reverse hash
(method REVERSE).

=head2 PREFIXES ( $uri )

Get all known prefixes of a namespace URI in sorted order.

=head2 REVERSE

Calling C<< $ns->REVERSE >> is equal to C<< RDF::SN->new($ns) >>. See
L<RDF::SN> for details.

=head2 SELECT ( prefix[es] )

In list context, returns a sorted list of prefix-namespace pairs, which
can be used to assign to a hash. In scalar context, returns the namespace
of the first prefix that was found. Prefixes can be passed as single arguments
or separated by commas, vertical bars, and spaces.

=head1 INTERNAL METHODS

=head2 SET ( $prefix => $namespaces [, $warn ] )

Set or add a namespace mapping. Errors are ignored unless enabled as warnings
with the third argument. Returns true if the mapping was successfully added.

=head2 MAP ( $code [, prefix[es] ] )

Internally used to map particular or all prefixes. Prefixes can be selected as
single arguments or separated by commas, vertical bars, and spaces. In scalar
context, C<$_> is set to the first existing prefix (if found) and C<$code> is
called. In list context, found prefixes are sorted at mapped with C<$code>.

=head2 GET ( $uri )

This method is used internally to create URIs as return value of the URI
method and all lowercase shortcut methods, such as C<foaf_Person>. By default
it just returns C<$uri> unmodified.

=head1 SEE ALSO

There are several other CPAN modules to deal with IRI namespaces, for instance
L<RDF::Trine::Namespace>, L<RDF::Trine::NamespaceMap>, L<URI::NamespaceMap>,
L<RDF::Prefixes>, L<RDF::Simple::NS>, L<RDF::RDFa::Parser::Profile::PrefixCC>,
L<Class::RDF::NS>, L<XML::Namespace>, L<XML::CommonNS> etc.

=encoding utf8

=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2013- by Jakob Voß.
 
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
